src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_tty.hh"
+
+#include "base/trace.hh"
+#include "debug/LupioTTY.hh"
+#include "dev/platform.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTTY.hh"
+
+#define LUPIO_TTY_INVAL    0x80000000
+
+/* Same fields for CTRL and STAT registers */
+#define LUPIO_TTY_WBIT  (1 << 0)
+#define LUPIO_TTY_RBIT  (1 << 1)
+
+namespace gem5
+{
+
+LupioTTY::LupioTTY(const Params &params)
+    : BasicPioDevice(params, params.pio_size),
+    writChar(-1), readChar(-1), writIntrEn(false), readIntrEn(false),
+    device(params.device), platform(params.platform)
+{
+    // setup serial device callbacks
+    device->regInterfaceCallback([this]() { dataAvailable(); });
+}
+
+void
+LupioTTY::dataAvailable()
+{
+      // unsure about these checks
+      // if interrupt enable bit is set and the read register is
+      // empty, then can raise an interrupt to receive data
+      if (readIntrEn && readChar == -1) {
+          DPRINTF(LupioTTY, "Data available, interrupting\n");
+          platform->postConsoleInt();
+     }
+}
+
+/*
+ * IRQ management
+ */
+void
+LupioTTY::lupioTTYUpdateIRQ()
+{
+    unsigned int irq;
+
+    irq = (writIntrEn && writChar != -1)
+        || (readIntrEn && readChar != -1);
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
```

Comment:
I don't understand how this can possibly be "generic" between ISAs right now. The platform is ISA (or really *board*) specific. The generic platform (`src/dev/platform.hh`) only has a `postConsoleInterrupt` and a `postPciInt`. I don't know how you're supposed to use these functions.

Reply:
The generic platform does seem to have a virtual function called `postConsoleInt()` (https://github.com/darchr/gem5/blob/develop/src/dev/platform.hh#L62). So I guess calling `platform->postConsoleInt()` should work regardless of the platform, no?

Reply:
My confusion is which console should this post the interrupt to? What if there are two consoles (i.e., in this case with both the uart and the lupioTTY

Reply:
I guess the current model doesn't allow this scenario. If you wanted to allow it, you'd probably need for `postConsoleInt()` to accept an ID of some sort to indicate which console is raising its IRQ.

That said, the lupio-TTY is replacing the uart now, so we actually don't need the support for two consoles.